File: frontend/types/product.ts
export type ProductStatus = "active" | "inactive";

export interface ProductCreate {
  name: string;
  category_id?: string | null;
  brand_id?: string | null;
  unit: string;
  units_per_box: number;
  loading_capacity: number;
  mrp: number;
  selling_price: number;
  gst_rate: number;
  minimum_stock: number;
  image?: string | null;
}

export type ProductUpdate = Partial<ProductCreate>;

export interface ProductResponse {
  id: string;
  sku: string;
  name: string;
  category_id: string | null;
  brand_id: string | null;
  unit: string;
  units_per_box: number;
  loading_capacity: number;
  mrp: number;
  selling_price: number;
  gst_rate: number;
  minimum_stock: number;
  image: string | null;
  status: ProductStatus;
  created_at: string;
}

// What GET /products returns — the catalog view. Price is pre-resolved
// server-side (customer's own price-list price, or base selling price for
// staff tokens) as `effective_price`.
export interface ProductCatalogResponse {
  id: string;
  sku: string;
  name: string;
  category_id: string | null;
  brand_id: string | null;
  unit: string;
  units_per_box: number;
  loading_capacity: number;
  mrp: number;
  effective_price: number;
  gst_rate: number;
  image: string | null;
}

// What GET /saved-products returns — a customer's bookmarked products, newest
// saved first, each carrying the full catalog view of the underlying product.
export interface SavedProductResponse {
  id: string;
  product: ProductCatalogResponse;
  created_at: string;
}

File: backend/app/schemas/product.py
import uuid
from datetime import datetime
from decimal import Decimal
from typing import Optional

from pydantic import BaseModel, ConfigDict

from app.core.enums import ProductStatus


class ProductCreate(BaseModel):
    name: str
    category_id: Optional[uuid.UUID] = None
    brand_id: Optional[uuid.UUID] = None
    unit: str
    units_per_box: int = 1
    mrp: Decimal
    selling_price: Decimal
    gst_rate: Decimal = Decimal("0")
    minimum_stock: int
    image: Optional[str] = None


class ProductUpdate(BaseModel):
    name: Optional[str] = None
    category_id: Optional[uuid.UUID] = None
    brand_id: Optional[uuid.UUID] = None
    unit: Optional[str] = None
    units_per_box: Optional[int] = None
    mrp: Optional[Decimal] = None
    selling_price: Optional[Decimal] = None
    gst_rate: Optional[Decimal] = None
    minimum_stock: Optional[int] = None
    image: Optional[str] = None


class ProductStatusUpdate(BaseModel):
    status: ProductStatus


class ProductResponse(BaseModel):
    id: uuid.UUID
    sku: str
    name: str
    category_id: Optional[uuid.UUID]
    brand_id: Optional[uuid.UUID]
    unit: str
    units_per_box: int
    mrp: Decimal
    selling_price: Decimal
    gst_rate: Decimal
    minimum_stock: int
    image: Optional[str]
    status: ProductStatus
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)


class ProductCatalogResponse(BaseModel):
    id: uuid.UUID
    sku: str
    name: str
    category_id: Optional[uuid.UUID]
    brand_id: Optional[uuid.UUID]
    unit: str
    units_per_box: int
    mrp: Decimal
    effective_price: Decimal
    gst_rate: Decimal
    image: Optional[str]

    model_config = ConfigDict(from_attributes=True)


class ProductDeleteResponse(BaseModel):
    id: uuid.UUID
    deleted_at: datetime

    model_config = ConfigDict(from_attributes=True)
